pennylane/transforms/split_non_commuting.py

```diff
@@ -316,73 +465,78 @@ def circuit(x):
+    for group_idx, obs_indices in enumerate(obs.grouping_indices):
+        mps_group = []
+        coeffs_group = []
+        idx_in_group = 0
 
-    # The number of measurements in each group
-    group_sizes = []
+        for idx in obs_indices:
+            obs = obs_list[idx]
+            coeff = coeff_list[idx]
 
-    # obs.grouping_indices is a list of lists, where each list contains the indices of
-    # observables that belong in each group.
-    for group_idx, obs_indices in enumerate(obs.grouping_indices):
-        mp_group = []
-        group_size = 0
-        for obs_idx in obs_indices:
-            # Do not measure Identity terms, but track their contribution with the offset.
-            if isinstance(obs_list[obs_idx], qml.Identity):
-                offset += coeffs[obs_idx]
+            if isinstance(obs, qml.Identity):
+                offset += coeff
             else:
-                new_mp = qml.expval(obs_list[obs_idx])
-                if new_mp in single_term_obs_mps:
-                    # If the Sum contains duplicate observables, it can be reused,
-                    # and the coefficients for each duplicate should be combined.
-                    single_term_obs_mps[new_mp] = (
-                        single_term_obs_mps[new_mp][0],
-                        [single_term_obs_mps[new_mp][1][0] + coeffs[obs_idx]],
-                        single_term_obs_mps[new_mp][2],
-                        single_term_obs_mps[new_mp][3],
-                    )
+                mp = qml.expval(obs)
+                if mp in single_term_obs_mps:
+                    single_term_obs_mps[mp].coeffs[0] += coeff
                 else:
-                    mp_group.append(new_mp)
-                    single_term_obs_mps[new_mp] = (
-                        [0],
-                        [coeffs[obs_idx]],
-                        group_idx,
-                        group_size,  # the index of this measurement in the group
+                    single_term_obs_mps[mp] = SingleTermMP(
+                        indices=[0],
+                        coeffs=[coeff],
+                        group_idx=group_idx,
+                        idx_in_group=idx_in_group,
                     )
-                    group_size += 1
+                    mps_group.append(mp)
+                    coeffs_group.append(coeff)
+                    idx_in_group += 1
 
-        if group_size > 0:
-            mp_groups.append(mp_group)
-            group_sizes.append(group_size)
+        if mps_group:
+            mps_groups.append(mps_group)
+            coeffs_per_group.append(coeffs_group)
+
+    if tape.shots.total_shots is not None and shot_dist_fn is not None:
+        shots_per_group = shot_dist_fn(tape.shots.total_shots, coeffs_per_group, seed)
+        tapes = [
+            tape.copy(measurements=mps, shots=int(shots))
+            for mps, shots in zip(mps_groups, shots_per_group)
+        ]
```

Comment:
```suggestion
        tapes = [
            tape.copy(measurements=mps, shots=int(shots))
            for mps, shots in zip(mps_groups, shots_per_group)
            if int(shots) != 0
        ]
```
We should allow a return of `shots=0` to correspond to dropping measurement of that term.

For example, if we had `H = 0.01*qml.X(0) + qml.Y(0)`, we can basically drop the `X` term.